theano/compile/builders.py

```diff
@@ -216,64 +256,94 @@ def __str__(self):
+        local_inputs = self.local_inputs
+        local_outputs = self.local_outputs
+        inp_len = len(local_inputs)
+        grad_op = self._grad_op
+
+        if isinstance(grad_op, OpFromGraph):
             self._grad_op_is_cached = True
+            self._grad_op_overrides_l = [None] * len(self.local_inputs)
             return
-        output_grads = [out_t() for out_t in self.output_types]
-        if self._grad_op is None:
-            self._grad_op = []
 
-        # we need to convert a list/function into an OfG instance
-        if isinstance(self._grad_op, list):
+        output_grads = [out_t() for out_t in self.output_types]
+        TYPE_ERR_MSG = 'Gradient override should be (single or list of)' \
+            'OpFromGraph | Ellipsis | None | 0 | callable, got %s'
+        # we need to convert _grad_op into an OfG instance
+        if grad_op is Ellipsis:
+            self._grad_op_tflags = bytes(inp_len)
+            all_grads_l = theano.gradient.grad(
+                cost=None,
+                known_grads=OrderedDict(izip(local_outputs, output_grads)),
+                wrt=local_inputs,
+                disconnected_inputs='ignore')
+            all_grads_ov_l = [None] * inp_len
+        elif grad_op is None:
+            all_grads_l = [inp.zeros_like() for inp in local_inputs]
+            all_grads_ov_l = [self.ofg_null_t()] * inp_len
+        elif grad_op is 0:
+            all_grads_l = [inp.zeros_like() for inp in local_inputs]
+            all_grads_ov_l = [self.ofg_discon_t()] * inp_len
+        elif isinstance(grad_op, list):
             goverrides_l = self._grad_op
-            if len(goverrides_l) > len(self.local_inputs):
+            if len(goverrides_l) != inp_len:
                 raise ValueError(
-                    'Can override %d gradients at most, got %d' % (
-                        len(self.local_inputs), len(goverrides_l)),
-                    self.goverrides_l)
-            if len(goverrides_l) < len(self.local_inputs):
-                goverrides_l += [None] * (
-                    len(self.local_inputs) - len(goverrides_l))
-            wrt_l = [lin for lin, gov in
-                     izip(self.local_inputs, goverrides_l) if not gov]
+                    'Need to override %d gradients, got %d' % (
+                        inp_len, len(goverrides_l)), goverrides_l)
             # compute non-overriding downsteam grads from upstreams grads
             # it's normal some input may be disconnected, thus the 'ignore'
+            wrt_l = [lin for lin, gov in izip(
+                self.local_inputs, goverrides_l) if gov is Ellipsis]
             gdefaults = iter(theano.gradient.grad(
                 cost=None,
-                known_grads=dict(izip(self.local_outputs, output_grads)),
+                known_grads=OrderedDict(izip(self.local_outputs, output_grads)),
                 wrt=wrt_l,
                 disconnected_inputs='ignore') if wrt_l else [])
```

Comment:
This this work if we have a list with all/none Ellipsis? Can you make sure this corner case is tested?